libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs

```diff
@@ -0,0 +1,707 @@
+  EQL -> pure total
+  GTE -> do Positive m <- arbitrary; choose (total, total + m)
+  GTH -> do Positive m <- arbitrary; choose (total + 1, total + m)
+  LTH ->
+    if count + 1 > total - 1
+      then
+        errorMess
+          ( "find n : "
+              ++ show count
+              ++ " < n  &&  n < "
+              ++ show total
+              ++ "\n"
+              ++ "[ SumsTo Term(count="
+              ++ show count
+              ++ ") LTH Sums(total="
+              ++ show total
+              ++ ") ]"
+              ++ " Cannot be solved."
+          )
+          msgs
+      else choose (count + 1, total - 1) -- (\ n -> count < n && n < total)
+  LTE ->
+    if count + 1 > total + 1
+      then
+        errorMess
+          ( "find n : "
+              ++ show count
+              ++ " < n  &&  n <= "
+              ++ show total
+              ++ "\n"
+              ++ "[ SumsTo Term(count="
+              ++ show count
+              ++ ") LTE Sums(total="
+              ++ show total
+              ++ ") ]"
+              ++ " Cannot be solved."
+          )
+          msgs
+      else choose (count + 1, total) -- (\ n -> count < n && n <= total)
+  CondAny -> suchThatErr msgs arbitrary (\x -> x > count)
+  CondNever xs -> errorMess "adjustTotalForCond called with (NeverCond _)" (msgs ++ xs)
+
+-- ===========================================================================
+-- The Sums class, for summing a projected c (where Adds c) from some richer type
+
+class (Show x, Adds x) => Sums t x | t -> x where
+  getsum :: t -> x
+  genT :: [String] -> x -> Gen t
+
+instance GoodCrypto c => Sums (IndividualPoolStake c) Rational where
+  getsum (IndividualPoolStake r _) = r
+  genT _ r = IndividualPoolStake r <$> arbitrary
+
+instance (Reflect era) => Sums (TxOut era) Coin where
+  getsum (TxOut (Shelley _) t) = coin (t ^. Core.valueTxOutL)
+  getsum (TxOut (Allegra _) t) = coin (t ^. Core.valueTxOutL)
+  getsum (TxOut (Mary _) t) = coin (t ^. Core.valueTxOutL)
+  getsum (TxOut (Alonzo _) t) = coin (t ^. Core.valueTxOutL)
+  getsum (TxOut (Babbage _) t) = coin (t ^. Core.valueTxOutL)
+  getsum (TxOut (Conway _) t) = coin (t ^. Core.valueTxOutL)
+  genT _ cn = genTxOutX reify cn
+
+txOutCoinL :: (Core.EraTxOut era) => Lens' (Core.TxOut era) Coin
+txOutCoinL = lens (\x -> coin (x ^. Core.valueTxOutL)) (\x c -> x & Core.valueTxOutL .~ (modifyCoin (const c) (x ^. Core.valueTxOutL)))
+
+genTxOutX :: Proof era -> Coin -> Gen (TxOut era)
+genTxOutX proof coins = do
+  (TxOut p txout) <- genTxOut proof
+  case p of
+    Shelley _ -> pure $ TxOut p (txout & txOutCoinL .~ coins)
+    Allegra _ -> pure $ TxOut p (txout & txOutCoinL .~ coins)
+    Mary _ -> pure $ TxOut p (txout & txOutCoinL .~ coins)
+    Alonzo _ -> pure $ TxOut p (txout & txOutCoinL .~ coins)
+    Babbage _ -> pure $ TxOut p (txout & txOutCoinL .~ coins)
+    Conway _ -> pure $ TxOut p (txout & txOutCoinL .~ coins)
```

Comment:
This may be ugly, but it prevents polluting the types of about 100 functions with the Core.EraTxOut era constraint.
I remove the parenthesis.

Reply:
Note that I didn't add `EraTxOut`, I added `Reflect` constraint, which implies `EraTxOut`.  You said it yourself it is ok to use `Reflect` in generic tests. Also this is what I mentioned during the meeting today. If we can avoid pattern matching on `Proof` we should, because it will reduce amount of work we have to do during addition of each era.

Reply:
But, it does not supply the Arbitrary instance for the type family TxOut. 
We need (Arbitrary (TxOut era)), and the pattern match supplies that.

Reply:
Just give it a try and you will see that I am right. This is the part that takes care of `Abitrary`:
```
TxOut p txOut <- genTxOut proof
```

Reply:
OK only one use of the lens. done